File: util/src/math/point.rs
use std::ops::{Add, AddAssign};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Point2 {
    pub x: i32,
    pub y: i32,
}

impl Point2 {
    pub const fn new(x: i32, y: i32) -> Self {
        Self { x, y }
    }

    pub const fn zero() -> Self {
        Self::new(0, 0)
    }

    pub const fn is_negative(&self) -> bool {
        self.x.is_negative() || self.y.is_negative()
    }

    pub fn polar_angle_to(self, other: Self) -> f64 {
        let x_diff: f64 = (other.x - self.x) as f64;
        let y_diff: f64 = (other.y - self.y) as f64;

        f64::atan2(y_diff, x_diff)
    }

    pub fn manhattan_distance(self, to: Self) -> i32 {
        i32::abs(self.x - to.x) + i32::abs(self.y - to.y)
    }
}

impl Default for Point2 {
    fn default() -> Self {
        Point2::zero()
    }
}

impl Add for Point2 {
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        Self {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
        }
    }
}

impl Add<&Self> for Point2 {
    type Output = Self;

    fn add(self, rhs: &Self) -> Self::Output {
        Self {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Point3 {
    pub x: i32,
    pub y: i32,
    pub z: i32,
}

impl Point3 {
    pub fn new(x: i32, y: i32, z: i32) -> Self {
        Point3 { x, y, z }
    }

    pub fn zero() -> Self {
        Point3::new(0, 0, 0)
    }
}

impl Default for Point3 {
    fn default() -> Self {
        Point3::zero()
    }
}

impl Add for Point3 {
    type Output = Point3;

    fn add(self, rhs: Self) -> Self::Output {
        Point3::new(self.x + rhs.x, self.y + rhs.y, self.z + rhs.z)
    }
}

impl AddAssign for Point3 {
    fn add_assign(&mut self, other: Self) {
        self.x += other.x;
        self.y += other.y;
        self.z += other.z;
    }
}

File: Sources/Base/Point.swift
import Foundation

public struct Point2: Equatable, Hashable {
    public var x: Int
    public var y: Int

    public static var zero: Self {
        .init()
    }

    public var isNegative: Bool {
        x < 0 || y < 0
    }

    public init(
        x: Int = 0,
        y: Int = 0
    ) {
        self.x = x
        self.y = y
    }

    public func polarAngle(to other: Self) -> Double {
        let xDiff = Double(other.x - self.x)
        let yDiff = Double(other.y - self.y)

        return atan2(yDiff, xDiff)
    }

    public func manhattanDistance(to other: Self) -> Int {
        abs(self.x - other.x) + abs(self.y - other.y)
    }

    public func window() -> [Self] {
        [
            Self(x: x - 1, y: y - 1),
            Self(x: x, y: y - 1),
            Self(x: x + 1, y: y - 1),
            Self(x: x - 1, y: y),
            self,
            Self(x: x + 1, y: y),
            Self(x: x - 1, y: y + 1),
            Self(x: x, y: y + 1),
            Self(x: x + 1, y: y + 1),
        ]
    }

    public static func + (_ lhs: Self, _ rhs: Self) -> Self {
        Self(x: lhs.x + rhs.x, y: lhs.y + rhs.y)
    }

    public static func += (_ lhs: inout Self, _ rhs: Self) {
        lhs.x += rhs.x
        lhs.y += rhs.y
    }
}

public struct Point3: Equatable, Hashable {
    public var x: Int
    public var y: Int
    public var z: Int

    public static var zero: Self {
        .init()
    }

    public init(
        x: Int = 0,
        y: Int = 0,
        z: Int = 0
    ) {
        self.x = x
        self.y = y
        self.z = z
    }

    public func manhattanDistance(to other: Self) -> Int {
        abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z - other.z)
    }

    public static func + (_ lhs: Self, _ rhs: Self) -> Self {
        Self(x: lhs.x + rhs.x, y: lhs.y + rhs.y, z: lhs.z + rhs.z)
    }

    public static func += (_ lhs: inout Self, _ rhs: Self) {
        lhs.x += rhs.x
        lhs.y += rhs.y
        lhs.z += rhs.z
    }

    public static func - (_ lhs: Self, _ rhs: Self) -> Self {
        Self(x: lhs.x - rhs.x, y: lhs.y - rhs.y, z: lhs.z - rhs.z)
    }

    public static func -= (_ lhs: inout Self, _ rhs: Self) {
        lhs.x -= rhs.x
        lhs.y -= rhs.y
        lhs.z -= rhs.z
    }
}

public struct Size: Equatable, Hashable, Comparable {
    public var width: Int
    public var height: Int

    public init(
        width: Int,
        height: Int
    ) {
        self.width = width
        self.height = height
    }

    public static func < (lhs: Size, rhs: Size) -> Bool {
        lhs.width < rhs.width && lhs.height < rhs.height
    }
}
